### ISSUE TITLE
Implement depth buffer-based collision for GPUParticles

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#3898]
### Describe the project you are working on

*Related to https://github.com/godotengine/godot-proposals/issues/3865.*

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

- Godot 4.0 supports GPUParticles collision using pre-placed collider nodes. A real-time heightmap shape (which can be parented to a Camera3D) is also available, but it is not suited to all kinds of particle collisions – it's good for weather effects, though.
- Godot `3.x` does not support any form of collision in GPU-based particles.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a `use_collision` property to GPUParticles (`master`) and Particles (`3.x`) based on the depth buffer, using the approach described here: https://github.com/Wolfe2x7/3D-particle-occlusion-collision-for-Godot-3

Depth buffer-based collision has a few advantages:

- No manual collider setup is required.
- Works well indoors and outdoors.

It also has downsides:

- If a particle leaves the screen or is occluded by geometry, it will be considered to have collided with something (and will therefore disappear).

In the form linked above, this collision only supports disappearing on contact (and not bouncing). It may be possible to implement bouncing, but given the caveats of depth buffer-based collision, it may not be worth it.
Nonetheless, I think disappearing on contact is still valuable in many cases, such as weather effects in `3.x`. In `master`, this can also be useful if you don't want to manually place particle colliders in an indoor level.

Note that depth buffer-based collision isn't feasible on CPUParticles – or not at a reasonable performance level at least.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This is essentially copying the MIT-licensed code from https://github.com/Wolfe2x7/3D-particle-occlusion-collision-for-Godot-3 and exposing it as a property in the built-in ParticlesMaterial :slightly_smiling_face: 

In `3.x`, this will require one more shader version to be created when particle collision is enabled on at least one ParticlesMaterial. This will have a small performance cost, but only when the feature is used in the project.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can already be implemented with a custom shader as linked above.

### Is there a reason why this should be core and not an add-on in the asset library?

Given community feedback over the years, particle collision is quite commonly desired by users. While Godot 4.0 will provide a good implementation of particle collision, it still requires a bit of manual work which may be too high of a barrier of entry (especially in complex indoor levels). Also, many users will stick to `3.x` for their current 3D project.

**Comment by Wolfe2x7** [comment]
> In the form linked above, this collision only supports disappearing on contact (and not bouncing). It may be possible to implement bouncing, but given the caveats of depth buffer-based collision, it may not be worth it.

How about simple canned bouncing? With the shader, you can code reactions other culling the particle. In the project, the fountain water loses some Y velocity and gains a random XZ velocity when it strikes the ball. The tennis ball launcher has the ramp's normal in a uniform and uses reflect().

To demonstrate further, I looked into a flamethrower effect. Is a simple splash behavior like this worth it? :)
[https://user-images.githubusercontent.com/72348938/151734508-6e118af9-aed9-499b-9c24-65317c25b6e5.mp4](https://user-images.githubusercontent.com/72348938/151734508-6e118af9-aed9-499b-9c24-65317c25b6e5.mp4)

There are some potential basic reactions upon collision:
1. Cull the particle
2. Stop the particle, like goop that covers the wall until the particles expire
3. Snap the particle to the surface, for area effects akin to decals
4. Bounce around a provided normal vector (explicitly; not reflected), similar to the fountain
5. Reflect upon a provided normal vector, like the tennis balls
6. Splash perpendicular to the emission vector, like the flamethrower

I think reactions 2 and 3 are simple enough for serious consideration, and number 6 may be desirable. Number 4 and 5 beckon for more customization, so maybe they should just be left to personalized shaders.

**Comment by Calinou** [comment]
> I think reactions 2 and 3 are simple enough for serious consideration, and number 6 may be desirable. Number 4 and 5 beckon for more customization, so maybe they should just be left to personalized shaders.

I agree this makes sense to support if possible. I don't know if it will behave well when the camera moves though.

**Comment by Rodeo-McCabe** [comment]
I really like this idea.

>     6. Splash perpendicular to the emission vector, like the flamethrower

I think it should splash perpendicular to the collision normal, though. Being able to get collision information similar to KinematicCollision would allow for extensive customization easily through GDscript. But I don't know if that kind of thing is feasible for particles.

**Comment by Wolfe2x7** [comment]
> I think it should splash perpendicular to the collision normal, though. Being able to get collision information similar to KinematicCollision would allow for extensive customization easily through GDscript. But I don't know if that kind of thing is feasible for particles.

Properly incorporating normals with this effect would require [access to a normal buffer along with the depth buffer](https://github.com/godotengine/godot-proposals/issues/798). If I understand it correctly, you could then read the normal buffer at the same UV position as the depth buffer and use that for writing collisions. 👍

Without that, you could hypothetically prepare a normal map ahead of time for colliding with something static, but otherwise pretty much all the particle shader can do is know when a particle is occluded by a surface and enact simple effects.

Who knows what wizards could do with it, though. That's why I wanted to get the idea out there. :)

**Comment by unfa** [comment]
Depth and normal buffer based particle collision and bouncing I think is quite necessary, as it's a zero-setup function. It fits Godot's philosophy of things "just working" but allowing more options if desired.

Right now I think it's quite a lot of work to manually build particle colliders out of primitives and SDF fields for a large game level.

**Comment by Calinou** [comment]
The way particle collision works in Godot 4 is completely different, so I don't think depth buffer-based collision is a viable approach there (with all its limitations on top of that).

What we should do instead is provide import hints to perform particle collider generation on import as well as options in the Advanced Import Settings dialog.
